DUMB 2.0 and 0.9.3 support

SimonN opened the pull request:
Specifying `-lm` on Arch Linux is necessary for DUMB support: Otherwise, cmake concludes that DUMB is not available. I used the DUMB 1.0-1 binary package from the Arch repo. This Arch package points to [kode54's fork of DUMB](https://github.com/kode54/dumb) with version 1.0 as suggested by the DUMB homepage, **not to version 0.9.3 that A5 officially supports.**

I didn't test whether this is required on other systems, or whether it breaks the build on other systems.

**Still no successful playback:** I've merely fixed A5's build for me. I can load a module with Allegro 5, but then my application will immediately crash within DUMB. Crashlog follows. I haven't debugged this crash yet! The fault could be in DUMB, in the DUMB Arch package, in A5, in my code, in a faulty module, ...

I'll understand if you won't merge this PR before we have seen successful playback with DUMB 1.0 and A5. According to @SiegeLord, DUMB 1.0 made breaking changes. Still, I'm opening this PR to share my findings, allowing others to shed more light on this.

```
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff3dd1962 in ?? () from /usr/lib/libdumb.so
#2  0x00007ffff3dd2d15 in dumb_read_it_quick () from /usr/lib/libdumb.so
#3  0x00007ffff3dd0996 in dumb_read_it () from /usr/lib/libdumb.so
#4  0x00007ffff78d5d74 in mod_stream_init (f=0x555556284600, buffer_count=3,
    samples=2048, loader=0x7ffff3dd0990 <dumb_read_it>)
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:213
#5  0x00007ffff78d625c in _al_load_it_audio_stream (filename=<optimized out>,
    buffer_count=3, samples=2048)
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:391
```

For reference, here's an excerpt from cmake's error log when DUMB's libraries don't specify `-lm`:

```
Run Build Command:"/usr/bin/make" "cmTC_5d751/fast"
/usr/bin/make -f CMakeFiles/cmTC_5d751.dir/build.make CMakeFiles/cmTC_5d751.dir/build
make[1]: Entering directory '/home/simon/c/notown/allegro5/simon/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_5d751.dir/src.c.o
/usr/bin/cc   -msse -W -Wall -Wpointer-arith -Wmissing-declarations -Wstrict-prototypes -Wmissing-prototypes -DDUMB_COMPILES   -o CMakeFiles/cmTC_5d751.dir/src.c.o   -c /home/simon/c/notown/allegro5/simon/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_5d751
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_5d751.dir/link.txt --verbose=1
/usr/bin/cc  -msse -W -Wall -Wpointer-arith -Wmissing-declarations -Wstrict-prototypes -Wmissing-prototypes -DDUMB_COMPILES    -rdynamic CMakeFiles/cmTC_5d751.dir/src.c.o  -o cmTC_5d751 /usr/lib/libdumb.so 
/usr/lib/libdumb.so: undefined reference to `__pow_finite'
/usr/lib/libdumb.so: undefined reference to `__log_finite'
/usr/lib/libdumb.so: undefined reference to `__exp_finite'
/usr/lib/libdumb.so: undefined reference to `sin'
/usr/lib/libdumb.so: undefined reference to `__log2_finite'
/usr/lib/libdumb.so: undefined reference to `cos'
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_5d751.dir/build.make:99: cmTC_5d751] Error 1
make[1]: Leaving directory '/home/simon/c/notown/allegro5/simon/CMakeFiles/CMakeTmp'
make: *** [Makefile:126: cmTC_5d751/fast] Error 2

Source file was:

            #include <dumb.h>
            int main(void)
            {
                dumb_register_stdfiles();
                return 0;
            }
```

SimonN commented:
More light on the crash inside DUMB 1.0 A5 runtime! I've built DUMB 1.0 with debugging flags, let A5 use that debugging build, and get a much more informative stacktrace.

Explanation: `siege.c` is SiegeLord's example A5 program from issue #670 that loads a music file given on the command line, then immediately plays it.

```
#0  0x00007ffff73318a0 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7332f09 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff732a0dc in __assert_fail_base () from /usr/lib/libc.so.6
#3  0x00007ffff732a153 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007ffff43afc93 in register_dumbfile_system (dfs=0x7ffff78be2c0 <dfs>)
    at /home/simon/c/notown/dumb/src/core/dumbfile.c:36
#5  0x00007ffff76ba0b7 in init_libdumb ()
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:346
#6  0x00007ffff76ba14f in init_libdumb ()
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:451
#7  _al_load_mod_audio_stream_f (f=f@entry=0x5555557c51e0, 
    buffer_count=buffer_count@entry=4, samples=samples@entry=2048)
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:452
#8  0x00007ffff76ba19c in _al_load_mod_audio_stream (filename=<optimized out>, 
    buffer_count=4, samples=2048)
    at /home/simon/c/notown/allegro5/addons/acodec/modaudio.c:368
#9  0x0000555555554acb in main (argc=2, argv=0x7fffffffe798) at siege.c:15
```

I suppose:
* The interface, or at least the calling conventions, between DUMB 0.9.3 and DUMB 1.0 differ.
* A5 calls the 1.0 interface in such a way that the DUMB assert fails.
* Either A5 or DUMB has an issue, provided A5 would like to offer support for DUMB 1.0.

Now it's time to dig deep into A5 source and the DUMB API...

SimonN commented:
I've fixed the API mismatches to the best of my knowledge. I've submitted new [documentation about `DUMBFILE_SYSTEM` to DUMB](https://github.com/kode54/dumb/blob/master/DUMBFILE_SYSTEM.md), this got merged today.

Ive changed A5's `modaudio.c` to call DUMB 1.0, but I'm not yet submitting my `modaudio.c` to this A5 pull request: A5 returns null streams even on XM files that it played with DUMB 1.0 before I started hacking. I want to more work on this first. (If you want to see my changes, they're in my public github fork of allegro5, but I rebase public branches heavily until I finally make PRs from them.)

I'll continue to investigate. For any questions, I'll sit in `irc.freenode.net #allegro` or `irc.quakenet.org #lix`, or reply here.

SimonN commented:
I've got it working! XM, IT, MOD, S3M all play well with the DUMB git HEAD.

DUMB supports even more tracked formats in 1.0, but Allegro doesn't register loaders for them yet. We can look into that soon.

My code relies on `dumb.h` reporting the correct version. Since the DUMB commit tagged 1.0 reports 0.9.3 internally, I've proposed upstream to release another tagged 1.y.z > 1.0.0: https://github.com/kode54/dumb/issues/48

beoran commented:
As an aside, I am glad this is coming together! Thank you very much, Simon,
for your relentless efforts.

On 17 Sep 2017 10:37, "Simon Na." <notifications@github.com> wrote:

> I've got it working! XM, IT, MOD, S3M all play well with the DUMB git HEAD.
>
> DUMB supports even more tracked formats in 1.0, but Allegro doesn't
> register loaders for them yet. We can look into that soon.
>
> My code relies on dumb.h reporting the correct version. Since the DUMB
> commit tagged 1.0 reports 0.9.3 internally, I've proposed upstream to
> release another tagged 1.y.z > 1.0.0: kode54/dumb#48
> <https://github.com/kode54/dumb/issues/48>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/liballeg/allegro5/pull/823#issuecomment-330029659>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEWeazhmdM_FSV3mxlghReLJjLLvUNmks5sjNpMgaJpZM4PVl9N>
> .
>

SimonN commented:
You're welcome!

Feels so good when research and work pay off. I haven't contributed to either library before and dug through internals in both.

SimonN commented:
News from DUMB upstream: We're working on stabilizing the API for DUMB 2.0. kode54, the maintainer, hasn't planned any breaking changes, only bugfixes. The DUMB header needs small clean-up in the public API, and I'm helping upstream getting all the ducks in a row.

This means **A5 should wait until DUMB 2.0** before merging this PR. I estimate we'll have DUMB 2.0 in 2 weeks. Then, I'll push another commit to this PR's target branch to make A5 check for DUMB 2.0 instead of 1.0.

DUMB 2.0 has a catch-all loader to guess file format from an open file. Looks like I can simplify the A5 code and allow all the 10+ tracker formats in one go.

SiegeLord commented on addons/acodec/acodec.c:
```diff
@@ -38,6 +38,48 @@ bool al_init_acodec_addon(void)
 #endif
 
 #ifdef ALLEGRO_CFG_ACODEC_MODAUDIO
+#include <dumb.h>
```
Probably shouldn't include a header inside the body of a function.

SimonN replied:
Good catch, thanks.

SimonN commented:
I expect the DUMB 2.0 API to stabilize within a week. I have pushed a commit to this PR (A5-DUMB-2.0-interop) that uses the most likely DUMB 2.0 API, and still should build with DUMB 0.9.3.

Once DUMB 2.0 is tagged upstream, I'll rebase this PR onto A5 master and squash commits as appropriate.

I'd love to make it onto the A5 October release before the ship sails!

beoran commented:
I am not the one to decide but I consider this feature important enough to
delay the next Allegro 5 release for. The DUMB dependency was getting
stale, upgrading to 2.0 seems a great idea.

On 21 Sep 2017 11:04 a.m., "Simon Naarmann" <notifications@github.com>
wrote:

> I expect the DUMB 2.0 API to stabilize within a week. I have pushed a
> commit to this PR (A5-DUMB-2.0-interop) that uses the most likely DUMB 2.0
> API, and still should build with DUMB 0.9.3.
>
> Once DUMB 2.0 is tagged upstream, I'll rebase this PR onto A5 master and
> squash commits as appropriate.
>
> I'd love to make it onto the A5 October release before the ship sails!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/liballeg/allegro5/pull/823#issuecomment-331097407>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEWeQNIvONau1RRy6-1RagSJC1yfVTMks5skiaegaJpZM4PVl9N>
> .
>

SiegeLord commented:
It's fine to change MOD_FILE's declaration to whatever you need it to be, it's purely an internal implementation detail with no user-visible API or ABI effects.

SimonN commented:
If it's purely internal, I'll keep `MOD_FILE` as-is, even if it might fail subtly for gigantic module files of several hundred GB.

We have gold-plated the DUMB API for 64-bit file offsets (module files > 2 GB), but that's only because we want the API future-proof. Maybe that was unnecessary already.

SimonN commented:
DUMB 2.0 is tagged upstream! @SiegeLord will look over this PR one more time tomorrow, afterwards I'll rebase and squash.

SiegeLord commented on addons/acodec/acodec.c:
```diff
@@ -38,6 +41,47 @@ bool al_init_acodec_addon(void)
+   ret &= al_register_audio_stream_loader_f(".mod", _al_load_dumb_audio_stream_f);
+   ret &= al_register_audio_stream_loader(".mtm", _al_load_dumb_audio_stream);
+   ret &= al_register_audio_stream_loader_f(".mtm", _al_load_dumb_audio_stream_f);
+   ret &= al_register_audio_stream_loader(".nst", _al_load_dumb_audio_stream);
```
Didn't you just say it wasn't listed?

fatcerberus replied:
I think he meant it wasn't listed in the DUMB documentation.

SimonN replied:
Earlier upstream discussion: https://github.com/kode54/dumb/issues/53#issuecomment-330127761 -- based on that, I decided to add `.nst` in the code above.

I'm considering to remove `.nst` above because, after googling around, it seems really hard to find NST modules with that extension. Not even [modarchive](https://modarchive.org/index.php) lists NST it in their format choice.

SiegeLord commented:
I tried this with dumb2, and found one issue. It appears that the stream length is not reported correctly (e.g. I used `ex_stream_seek`, and it did not display the length). This seems to work fine with dumb1 (with this code). Maybe it's a dumb2 issue though.

SimonN commented:
> stream length is not reported correctly (e.g. I used ex_stream_seek, and it did not display the length). This seems to work fine with dumb1 (with this code).

Excellent catch. Can reproduce exactly as you told me: With my latest A5 PR and `ex_stream_seek hybridsong220.xm`, I can seek in the earlier dumb cbbaa56733f7 from 2 weeks ago, but not in dumb 2.0.0. Will examine!

SimonN commented:
Fixed both issues (`.nst` registration and failure to seek). Rebased onto master, but not yet squashed.

I fixed the seek failure with `dumb_read_any` instead of `dumb_read_any_quick`.

Strangely, `dumb_read_any_quick` worked 20 commits ago. I'm confused and will investigate, maybe our non-perfect DUMB API 2.0 64-bit conversion is at fault. But even if I find the real cause, using `dumb_read_any_quick` would be a minor performance gain merely, not a massive feature blocker.

SimonN commented:
I've bisected the DUMB codebase, and now I feel dumb: In this A5 PR, I `#if` the DUMB header version. For >= 2.0, I used `dumb_read_any_quick`; for smaller DUMB, I used the old dumb readers. That's why my PR behaves so differently with DUMB 2.0 and 1.0.

I should always use `dumb_read_any` to allow seeking, never `dumb_read_any_quick`. It's fixed in the above a947cb2.

The 64-bit DUMB API has nothing to do with this, sorry for tying @SiegeLord into an hour of rambling!

SiegeLord commented:
Ok, tested this again and it seemed to work fine. Feel free to squash it when convenient.

beoran commented:
@SiegeLord On the subject of squashing, I see that now Github has a convenient looking "squash and merge" button for Allegro developers with sufficinet privileged. Do you think it is OK to actually use this button once all reviews and tests are done? It saves the contributor the effort of having to do a manual squash.

SimonN commented:
Squashed into 2 commits: Fixing whitespace, then a clean diff of the feature.

SiegeLord commented:
@beoran, it doesn't do the rebase, as far as I can tell. I really don't like having merges in the history. We very rarely ask people to squash and rebase (I always rebase myself before merging), Simon is just being super nice :D.

elias-pschernig commented:
I hate repositories who have 1000ds of useless merge commits - so I'm really glad we avoid them in Allegro (except for actual branch merges, which are rare now that we switched to a single development branch).

beoran commented:
I see, I can agree with that. I will refrain from using the button, then.

On 28 Sep 2017 20:14, "elias-pschernig" <notifications@github.com> wrote:

> I hate repositories who have 1000ds of useless merge commits - so I'm
> really glad we avoid them in Allegro (except for actual branch merges,
> which are rare now that we switched to a single development branch).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/liballeg/allegro5/pull/823#issuecomment-332920043>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEWeY-rAGzpHOiE6xJsVy_XtQjBQTYpks5sm-H0gaJpZM4PVl9N>
> .
>

SiegeLord commented:
Alright, merged! Thanks a lot Simon, you're the best!

SimonN commented:
You're welcome! Thanks to SiegeLord for the frequent reviews, and to everyone for feedback and encouragement.
